### PR TITLE
[MS-51] Use file seeking in `Membrane.WAV.Serializer`

### PR DESCRIPTION
**Pull request by balins**
If `Membrane.File.SeekEvent` is available, `Membrane.WAV.Serializer` will emit event that updates invalid WAV header. Otherwise, `Membrane.WAV.Serializer` works like before.

### CODE REVIEW

File: test/serializer/serializer_test.exs [hunk 1]
```diff
@@ -101,5 +102,38 @@ defmodule Membrane.WAV.SerializerTest do
       assert_sink_buffer(pid, :sink, %Buffer{payload: ^payload})
       Pipeline.stop_and_terminate(pid, blocking?: true)
     end
+
+    @tag :tmp_dir
+    test "create valid file when `update_header?` = `true`", %{tmp_dir: tmp_dir} do
```
Comment (bblaszkow06) [inline test/serializer/serializer_test.exs]: The markdown syntax doesn't make sense here as it will never go through any markdown processor:
```suggestion
    test "create valid file when 'update_header?' is enabled", %{tmp_dir: tmp_dir} do
```

File: README.md [hunk 2]
```diff
@@ -39,8 +39,9 @@ The Serializer adds WAV header to the raw audio in uncompressed, PCM format.
 
 ### Warning
 
-The Serializer creates header with invalid blocks which depend on the file size. To fix that, using
-`Membrane.WAV.Postprocessing` is recommended.
+The Serializer creates header with invalid blocks which depend on the file size. You can bypass this 
+using `Membrane.File.Sink` to save the file or fixing the header afterwards with `Membrane.WAV.Postprocessing`
+if you need to use any other sink.
```
Comment (bblaszkow06) [inline README.md:44]: This sounds like the element is broken and you can somehow work around the issue. I'd rephrase it to something mentioning, that if the element after serializer doesn't support seek events, you may work **this** around by disabling option in the element and using Postprocessing

File: lib/membrane_wav/serializer.ex [hunk 3]
```diff
@@ -32,6 +36,18 @@ defmodule Membrane.WAV.Serializer do
                 Used when converting demand from buffers into bytes.
                 """,
                 default: 2048
+              ],
+              update_header?: [
```
Comment (bblaszkow06) [inline lib/membrane_wav/serializer.ex]: To better communicate that this is not something usually done, I'd rename it to `disable_seeking` or `disable_header_update`.

File: lib/membrane_wav/serializer.ex [hunk 4]
```diff
@@ -50,25 +66,27 @@ defmodule Membrane.WAV.Serializer do
...
 
     {:ok, state}
   end
 
   @impl true
   def handle_caps(:input, caps, _context, state) do
     buffer = %Buffer{payload: create_header(caps)}
-    state = %{state | header_created: true}
-
+    state = %{state | header_length: byte_size(buffer.payload)}
```
Comment (bblaszkow06) [inline lib/membrane_wav/serializer.ex]: This might be the place to subtract the 8 bytes not included in header length

File: lib/membrane_wav/serializer.ex [hunk 5]
```diff
@@ -77,20 +95,33 @@ defmodule Membrane.WAV.Serializer do
...
     caps = context.pads.output.caps
     demand_size = Caps.frames_to_bytes(frames, caps) * buffers_count
     {{:ok, demand: {:input, demand_size}}, state}
   end
 
   @impl true
-  def handle_process(:input, buffer, _context, %{header_created: true} = state) do
-    {{:ok, buffer: {:output, buffer}, redemand: :output}, state}
+  def handle_process_list(:input, _buffers, _context, %{header_length: 0}) do
+    raise(RuntimeError, "buffers received before caps, so the header is not created yet")
```
Comment (bblaszkow06) [inline lib/membrane_wav/serializer.ex]: If not using any custom exception, use `raise/1`
```suggestion
    raise "Buffers received before caps, cannot create the header"
```

File: lib/membrane_wav/serializer.ex [hunk 6]
```diff
@@ -77,20 +95,33 @@ defmodule Membrane.WAV.Serializer do
...
-    raise(RuntimeError, "buffer received before caps, so the header is not created yet")
+  def handle_process_list(:input, buffers, _context, %{data_length: data_length} = state) do
+    state =
+      buffers
+      |> Enum.reduce(data_length, fn %Buffer{payload: payload}, acc ->
+        acc + byte_size(payload)
+      end)
+      |> then(&Map.put(state, :data_length, &1))
+
+    {{:ok, buffer: {:output, [buffers]}, redemand: :output}, state}
```
Comment (bblaszkow06) [inline lib/membrane_wav/serializer.ex]: `buffers` is a list, no need to wrap it
```suggestion
    {{:ok, buffer: {:output, buffers}, redemand: :output}, state}
```

File: lib/membrane_wav/serializer.ex [hunk 7]
```diff
@@ -77,20 +95,33 @@ defmodule Membrane.WAV.Serializer do
...
 
-  def handle_process(:input, _buffer, _context, %{header_created: false}) do
-    raise(RuntimeError, "buffer received before caps, so the header is not created yet")
+  def handle_process_list(:input, buffers, _context, %{data_length: data_length} = state) do
+    state =
+      buffers
+      |> Enum.reduce(data_length, fn %Buffer{payload: payload}, acc ->
+        acc + byte_size(payload)
+      end)
+      |> then(&Map.put(state, :data_length, &1))
```
Comment (bblaszkow06) [inline lib/membrane_wav/serializer.ex]: A bit of nitpicking, but I don't like when, while piping, the "subject" changes. I'd split that:
```suggestion
    data_length =
      buffers
      |> Enum.reduce(data_length, fn %Buffer{payload: payload}, acc ->
        acc + byte_size(payload)
      end)
    
    state = %{state | data_length: data_length}
```

File: lib/membrane_wav/serializer.ex [hunk 8]
```diff
@@ -115,4 +146,32 @@ defmodule Membrane.WAV.Serializer do
...
+        ]
+
+      {:error, reason} ->
+        require Membrane.Logger
+
+        Membrane.Logger.warn("""
+        Couldn't update WAV header as `Membrane.File.SeekEvent` module is not available (reason: #{inspect(reason)}).
+        You can use `Membrane.File.Sink` in your pipeline to correctly save the file or fix it later using
+        `Membrane.WAV.Postprocessing`.
+        """)
```
Comment (bblaszkow06) [inline lib/membrane_wav/serializer.ex]: This should not go just with a warning that can be missed. In particular, the file plugin is now `:test` only and will "disappear" if running in `MIX_ENV=dev`. This scenario (enabled update and no 'seek' event is available) should raise with an error suggesting to disable the option and informing about consequences (i.e. the file won't have a valid header).

File: mix.exs [hunk 9]
```diff
@@ -36,7 +36,7 @@ defmodule Membrane.WAV.Plugin.Mixfile do
       {:ex_doc, "~> 0.26", only: :dev, runtime: false},
       {:dialyxir, "~> 1.1", only: :dev, runtime: false},
       {:credo, "~> 1.6", only: :dev, runtime: false},
-      {:membrane_file_plugin, "~> 0.7.0", only: :test},
+      {:membrane_file_plugin, "~> 0.8.0", only: :test},
```
Comment (bblaszkow06) [inline mix.exs]: If it's not available in 'dev', you won't be able to test it from `iex`. We can solve it by making the plugin an [`optional: true`](https://hexdocs.pm/mix/Mix.Tasks.Deps.html#module-dependency-definition-options) dependency

File: lib/membrane_wav/serializer.ex [hunk 10]
```diff
@@ -115,4 +146,32 @@ defmodule Membrane.WAV.Serializer do
...
+
+        []
+    end
+  end
```
Comment (bblaszkow06) [inline lib/membrane_wav/serializer.ex]: This is invoked from an app, so `ensure_loaded?` should be enough. Also, if the check was moved outside, you could use the proper struct creation syntax and have compile-time check for the keys:

```suggestion
  if Code.ensure_loaded?(Membrane.File.SeekEvent) do
    defp maybe_update_header_actions(%{
           update_header?: true,
           header_length: header_length,
           data_length: data_length
         }) do
      # subtracting 8 bytes as `file_length` field doesn't include "RIFF" header and the field itself
      file_length = header_length + data_length - 8

      [
        event: {:output, %Membrane.File.SeekEvent{position: @file_length_offset}},
        buffer: {:output, %Buffer{payload: <<file_length::32-little>>}},
        event: {:output, %Membrane.File.SeekEvent{position: @data_length_offset}},
        buffer: {:output, %Buffer{payload: <<data_length::32-little>>}}
      ]
    end
  else
    defp maybe_update_header_actions(%{update_header?: true}) do
      # Suppress warnings
      _offset = @file_length_offset
      _offset = @data_length_offset

      raise "Unable to update WAV file header using seek events. [...]"
    end
  end
```

This can be verified by replacing `position` with `positions` (assuming file plugin is still a test-only dependency) will result in compilation error for `test` env and clean compilation on `dev`

**Review by bblaszkow06 — CHANGES_REQUESTED** [review]

File: test/serializer/serializer_test.exs (hunk 1 above)
Comment (balins) [reply test/serializer/serializer_test.exs]: It was rather an emphasis for the reader that this is some \`variable\`. But surely can be removed if we don't follow such convention.

File: lib/membrane_wav/serializer.ex (hunk 10 above)
Comment (balins) [reply lib/membrane_wav/serializer.ex]: Surely that's way better!

File: test/serializer/serializer_test.exs [hunk 11]
```diff
@@ -92,26 +94,28 @@ defmodule Membrane.WAV.SerializerTest do
...
       assert_sink_buffer(pid, :sink, %Buffer{payload: ^payload})
+      refute_sink_event(pid, :sink, %Membrane.File.SeekEvent{})
+      assert_end_of_stream(pid, :sink)
+
       Pipeline.stop_and_terminate(pid, blocking?: true)
     end
 
     @tag :tmp_dir
-    test "create valid file when `update_header?` = `true`", %{tmp_dir: tmp_dir} do
+    test "create valid file when seeking is enabled", %{tmp_dir: tmp_dir} do
```
Comment (bblaszkow06) [inline test/serializer/serializer_test.exs:111]: That's even better than my suggestion :+1:

**Review by bblaszkow06 — APPROVED** [review]
cool
